runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/ActiveWorkRefresher.java

```diff
@@ -115,32 +132,72 @@ private void invalidateStuckCommits() {
     }
   }
 
+  /** Create {@link Heartbeats} and group them by {@link HeartbeatSender}. */
   private void refreshActiveWork() {
     Instant refreshDeadline = clock.get().minus(Duration.millis(activeWorkRefreshPeriodMillis));
+    Map<HeartbeatSender, Heartbeats> heartbeatsBySender =
+        aggregateHeartbeatsBySender(refreshDeadline);
 
+    if (heartbeatsBySender.isEmpty()) {
+      return;
+    }
+
+    if (heartbeatsBySender.size() == 1) {
+      // If there is a single HeartbeatSender, just use the calling thread to send heartbeats.
+      Map.Entry<HeartbeatSender, Heartbeats> heartbeat =
+          Iterables.getOnlyElement(heartbeatsBySender.entrySet());
+      sendHeartbeat(heartbeat);
+    } else {
+      // If there are multiple HeartbeatSenders, send out the heartbeats in parallel using the
+      // fanOutActiveWorkRefreshExecutor.
+      List<CompletableFuture<Void>> fanOutRefreshActiveWork = new ArrayList<>();
+      for (Map.Entry<HeartbeatSender, Heartbeats> heartbeat : heartbeatsBySender.entrySet()) {
+        fanOutRefreshActiveWork.add(
+            CompletableFuture.runAsync(
+                () -> sendHeartbeat(heartbeat), fanOutActiveWorkRefreshExecutor));
+      }
+
+      // Don't block until we kick off all the refresh active work RPCs.
+      @SuppressWarnings("rawtypes")
+      CompletableFuture<Void> parallelFanOutRefreshActiveWork =
+          CompletableFuture.allOf(fanOutRefreshActiveWork.toArray(new CompletableFuture[0]));
+      parallelFanOutRefreshActiveWork.join();
+    }
+  }
+
+  /** Aggregate the heartbeats across computations by HeartbeatSender for correct fan out. */
+  private Map<HeartbeatSender, Heartbeats> aggregateHeartbeatsBySender(Instant refreshDeadline) {
     Map<HeartbeatSender, Heartbeats.Builder> heartbeatsBySender = new HashMap<>();
 
     // Aggregate the heartbeats across computations by HeartbeatSender for correct fan out.
     for (ComputationState computationState : computations.get()) {
       for (RefreshableWork work : computationState.getRefreshableWork(refreshDeadline)) {
         heartbeatsBySender
             .computeIfAbsent(work.heartbeatSender(), ignored -> Heartbeats.builder())
-            .addWork(work)
-            .addHeartbeatRequest(computationState.getComputationId(), createHeartbeatRequest(work));
+            .add(computationState.getComputationId(), work, sampler);
       }
     }
 
-    heartbeatSender.accept(
-        heartbeatsBySender.entrySet().stream()
-            .collect(toImmutableMap(Map.Entry::getKey, e -> e.getValue().build())));
+    return heartbeatsBySender.entrySet().stream()
+        .collect(toImmutableMap(Map.Entry::getKey, e -> e.getValue().build()));
+  }
+
+  private void sendHeartbeat(Map.Entry<HeartbeatSender, Heartbeats> heartbeat) {
```

Comment:
could add noException or something to method name to help see above that it is safe to call inline or on executor without handling.

Reply:
done